File: backend/lib/lan_https.py
"""One decision, shared by both launchers: is there a LAN HTTPS listener?

A browser hands out ``AudioContext.audioWorklet``, the microphone, Web MIDI,
the clipboard and ``crypto.subtle`` only in a SECURE CONTEXT -- ``https://`` or
``localhost``. A second computer opening theDAW at ``http://<lan-ip>:5173`` is
neither, so its EDIT tab dies on ``ctx.audioWorklet`` being undefined. The fix
is a second Vite listener, the same app over TLS, on
:data:`backend.ports.LAN_HTTPS_PORT`.

Two launchers start theDAW -- ``backend/_devstack.py`` (web mode) and
``electron-ui/main/index.ts`` (desktop mode) -- and they must agree about
whether that listener runs, on which port, and why not when it does not.
:func:`plan_lan_https` is that agreement: pure, so both the table of cases and
the exact wording of the reason are testable without a certificate, a network
interface or a subprocess. The launchers differ only in how they collect the
three inputs (:func:`resolve_plan` does it for Python; the desktop shell reads
the same plan as JSON from ``python -m backend.lib.lan_https --json``).

Nothing here starts, kills or probes anything, and no failure it meets is
fatal: the LAN listener is a convenience, and theDAW must launch without it.
"""

from __future__ import annotations

import argparse
import json
import logging
import os
import sys
from collections.abc import Mapping, Sequence
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Optional, Protocol

from backend import ports

log = logging.getLogger(__name__)

__all__ = [
    "ENV_CERT",
    "ENV_ENABLED",
    "ENV_KEY",
    "ENV_PORT",
    "LISTENER_ARGS",
    "LISTENER_COMMAND",
    "LanHttpsPlan",
    "blocking_reason",
    "detect_lan_ips",
    "lan_address",
    "listener_env",
    "listener_port",
    "plan_lan_https",
    "read_settings",
    "resolve_plan",
]

#: The environment the listener itself reads (``frontend/vite.lan.config.ts``).
#: These three names are the contract with that file; do not rename one side.
ENV_CERT = "theDAW_HTTPS_CERT"
ENV_KEY = "theDAW_HTTPS_KEY"
ENV_PORT = "theDAW_HTTPS_PORT"
#: An explicit yes/no that beats the stored setting, for a launcher wrapper or
#: a one-off run: ``theDAW_LAN_HTTPS=0`` turns the listener off for this launch.
ENV_ENABLED = "theDAW_LAN_HTTPS"

#: ``settings.json`` -> ``app.lan_https``. Absent means ON: the whole point is
#: that a second device works without anyone configuring anything.
SETTING_SECTION = "app"
SETTING_KEY = "lan_https"

#: How the listener is started, from ``frontend/``. ``npx`` because vite is a
#: frontend devDependency, not a global.
LISTENER_ARGS: tuple[str, ...] = ("vite", "--config", "vite.lan.config.ts")
LISTENER_COMMAND = " ".join(("npx", *LISTENER_ARGS))

_FALSE = frozenset({"0", "false", "no", "off"})
_TRUE = frozenset({"1", "true", "yes", "on"})


class CertPathsLike(Protocol):
    """The shape :func:`backend.lib.lan_cert.ensure_lan_cert` returns."""

    cert: Path
    key: Path


@dataclass(frozen=True)
class LanHttpsPlan:
    """What the launchers do about the LAN listener this launch.

    ``reason`` is set only when ``enabled`` is False, and it is written to be
    read by the person looking at the log: it says which of the four things
    was missing, not that "something" was.
    """

    enabled: bool
    port: int
    url: Optional[str] = None
    cert_paths: Optional[CertPathsLike] = None
    reason: Optional[str] = None

    def log_line(self) -> str:
        """The one line a launcher prints about the LAN listener."""
        if self.enabled and self.url:
            return f"LAN (https): {self.url}"
        return f"LAN (https): off - {self.reason or 'unavailable'}"

    def as_json(self) -> dict[str, Any]:
        """The plan as the desktop shell reads it (``--json``)."""
        return {
            "enabled": self.enabled,
            "port": self.port,
            "url": self.url,
            "cert": str(self.cert_paths.cert) if self.cert_paths else None,
            "key": str(self.cert_paths.key) if self.cert_paths else None,
            "reason": self.reason,
        }


def _flag(raw: object) -> Optional[bool]:
    """A yes/no setting, or None when the value says nothing.

    ``settings.json`` is hand-editable and the environment carries strings
    only, so ``false`` and ``"false"`` and ``"0"`` all have to mean off.
    """
    if isinstance(raw, bool):
        return raw
    if isinstance(raw, str):
        text = raw.strip().lower()
        if text in _FALSE:
            return False
        if text in _TRUE:
            return True
    return None


def listener_port(env: Mapping[str, str]) -> int:
    """The port the LAN listener uses: ``theDAW_HTTPS_PORT`` or the default.

    A value that is not a usable TCP port is ignored rather than obeyed --
    ``vite.lan.config.ts`` falls back to the same default, so a typo must not
    leave the launcher and the listener disagreeing about the address.
    """
    raw = (env.get(ENV_PORT) or "").strip()
    if raw:
        try:
            port = int(raw)
        except ValueError:
            port = 0
        if 0 < port < 65536:
            return port
    return ports.LAN_HTTPS_PORT


def lan_address(lan_ips: Sequence[str]) -> Optional[str]:
    """The address other devices are told to use: the first real one."""
    return next((ip for ip in lan_ips if ip), None)


def blocking_reason(
    settings: Mapping[str, Any],
    env: Mapping[str, str],
    lan_ips: Sequence[str],
) -> Optional[str]:
    """Why the LAN listener cannot run, *ignoring the certificate*, or None.

    Split out from :func:`plan_lan_https` so a launcher can ask the cheap
    question first: minting a certificate shells out to openssl, and a launch
    that is switched off or has no network must never pay for that.
    """
    override = _flag(env.get(ENV_ENABLED))
    if override is False:
        return f"turned off for this launch by {ENV_ENABLED}"

    if override is None:
        section = settings.get(SETTING_SECTION)
        stored = section.get(SETTING_KEY) if isinstance(section, Mapping) else None
        if _flag(stored) is False:
            return f"turned off in Settings ({SETTING_SECTION}.{SETTING_KEY})"

    if lan_address(lan_ips) is None:
        return (
            "no LAN address - connect this machine to Wi-Fi or Ethernet "
            "for other devices to reach it"
        )
    return None


def plan_lan_https(
    settings: Mapping[str, Any],
    env: Mapping[str, str],
    lan_ips: Sequence[str],
    cert: Optional[CertPathsLike],
) -> LanHttpsPlan:
    """Whether to start the LAN HTTPS listener, where, and why not.

    Pure. Three things have to be true, and each failure names itself:

    * the user has not turned it off (``theDAW_LAN_HTTPS``, then
      ``settings.app.lan_https``; absent means on),
    * this machine has a LAN address to be reached at,
    * a certificate exists for it.
    """
    port = listener_port(env)

    blocked = blocking_reason(settings, env, lan_ips)
    if blocked is not None:
        return LanHttpsPlan(enabled=False, port=port, reason=blocked)

    if cert is None:
        return LanHttpsPlan(
            enabled=False,
            port=port,
            reason=(
                "no certificate - install openssl (Git for Windows ships one) "
                "and restart theDAW"
            ),
        )

    return LanHttpsPlan(
        enabled=True,
        port=port,
        url=f"https://{lan_address(lan_ips)}:{port}",
        cert_paths=cert,
    )


def listener_env(plan: LanHttpsPlan, base: Mapping[str, str]) -> dict[str, str]:
    """The child environment for the LAN listener, on top of ``base``.

    ``base`` is the caller's already-sanitised environment --
    :func:`backend.lib.launch_token.child_env` in the backend's launcher -- so
    the desktop shell's launch token never reaches a vite process. Only the
    four names the listener actually reads are added.
    """
    if not plan.enabled or plan.cert_paths is None:
        raise ValueError("listener_env is for an enabled plan with a certificate")
    env = dict(base)
    env["ENABLE_HMR"] = "true"
    env[ENV_CERT] = str(plan.cert_paths.cert)
    env[ENV_KEY] = str(plan.cert_paths.key)
    env[ENV_PORT] = str(plan.port)
    return env


def detect_lan_ips() -> list[str]:
    """This machine's LAN IPv4 addresses, best effort, most useful first.

    Imported where it is used: ``backend.modules.vj.sidecar`` owns the
    UDP-connect detection, and a launcher that merely asks about HTTPS should
    not pay for that module at import time.
    """
    try:
        from backend.modules.vj.sidecar import detect_lan_ip
    except Exception as exc:  # pragma: no cover - the module is part of the app
        log.debug("lan-https: LAN address detection unavailable (%s)", exc)
        return []
    ip = detect_lan_ip()
    return [ip] if ip else []


def read_settings() -> dict[str, Any]:
    """``settings.json`` as it is on disk, or ``{}`` when it cannot be read.

    Deliberately the raw file rather than ``SettingsStore``: a launcher runs
    before the backend exists, and building a store would create the file and
    take its lock. An unreadable or half-written file means "no preference
    recorded", which is the default -- on.
    """
    try:
        from backend.modules.settings.store import default_settings_path

        raw = json.loads(Path(default_settings_path()).read_text(encoding="utf-8"))
    except Exception as exc:
        log.debug("lan-https: no readable settings (%s) - using the defaults", exc)
        return {}
    return raw if isinstance(raw, dict) else {}


def resolve_plan(env: Optional[Mapping[str, str]] = None) -> LanHttpsPlan:
    """The plan for this machine, right now: settings, address, certificate.

    Never raises. Minting the certificate is the only slow step and it is
    skipped entirely once the listener has been turned off or there is nobody
    to serve, so the common "no network" launch costs one UDP socket.
    """
    environment = os.environ if env is None else env
    settings = read_settings()
    lan_ips = detect_lan_ips()

    # Ask for a certificate only when everything else already says yes, so a
    # disabled or address-less launch never shells out to openssl.
    blocked = blocking_reason(settings, environment, lan_ips)
    if blocked is not None:
        return LanHttpsPlan(
            enabled=False, port=listener_port(environment), reason=blocked
        )

    cert: Optional[CertPathsLike] = None
    try:
        from backend.lib.lan_cert import ensure_lan_cert

        cert = ensure_lan_cert(list(lan_ips))
    except Exception as exc:
        log.warning(
            "lan-https: no certificate for the LAN (%s). theDAW still runs; "
            "other devices reach it over plain http, where browsers block "
            "audio, mic and MIDI.",
            exc,
        )
        cert = None
    return plan_lan_https(settings, environment, lan_ips, cert)


def _main(argv: Optional[list[str]] = None) -> int:
    """``python -m backend.lib.lan_https --json``.

    How the desktop shell asks the same question the Python launcher asks,
    without reimplementing the answer in TypeScript. Always exits 0 and always
    prints one JSON object: a shell that cannot read a plan must fall back to
    "no LAN listener", not to a crash dialog.
    """
    parser = argparse.ArgumentParser(
        prog="python -m backend.lib.lan_https",
        description="Report whether theDAW serves the LAN over HTTPS.",
    )
    parser.add_argument(
        "--json",
        action="store_true",
        help="print the plan as JSON (the desktop shell reads this)",
    )
    args = parser.parse_args(argv)

    try:
        plan = resolve_plan()
    except Exception as exc:
        plan = LanHttpsPlan(
            enabled=False,
            port=ports.LAN_HTTPS_PORT,
            reason=f"could not be worked out ({exc})",
        )
    if args.json:
        print(json.dumps(plan.as_json()))
    else:
        print(plan.log_line())
    return 0


if __name__ == "__main__":
    sys.exit(_main())

File: electron-ui/main/lanHttps.ts
// The desktop shell's half of the LAN HTTPS listener (batch 11, H2).
//
// A browser exposes AudioContext.audioWorklet, the microphone, Web MIDI, the
// clipboard and crypto.subtle only in a SECURE CONTEXT -- https:// or
// localhost. A second computer opening theDAW at http://<lan-ip>:5173 is
// neither, so its EDIT tab dies on `ctx.audioWorklet` being undefined. Desktop
// mode therefore starts a second Vite listener in dev, the same app over TLS,
// beside electron-vite's own http one.
//
// WHETHER it starts is not decided here. `backend/lib/lan_https.py` owns that
// decision for both launchers (the setting, the LAN address, the certificate),
// and this side reads its answer from `python -m backend.lib.lan_https --json`
// rather than reimplementing it in TypeScript, where it would drift. What lives
// here is only the pure shape of that exchange: parsing the plan, composing the
// log line, and building the child's command and environment.
//
// No Electron import, so these run under plain node/tsx -- see
// ./lanHttps.test.ts, the same arrangement as ./downloadNaming.ts.

/** The plan `python -m backend.lib.lan_https --json` prints. */
export interface LanHttpsPlan {
  enabled: boolean
  /** The TLS port, whether or not the listener is enabled. */
  port: number
  /** `https://<lan-ip>:<port>`. Present exactly when enabled. */
  url: string | null
  /** PEM certificate path. Present exactly when enabled. */
  cert: string | null
  /** PEM key path. Present exactly when enabled. */
  key: string | null
  /** The vite executable to run. Present exactly when enabled. */
  vite: string | null
  /** Why it is off. Present only when it is. */
  reason: string | null
}

/** What the listener is started WITH, after the binary, from `frontend/`.
 *  Mirrors `LISTENER_ARGS` in backend/lib/lan_https.py. */
const LISTENER_ARGS = ['--config', 'vite.lan.config.ts']

/** The environment names `frontend/vite.lan.config.ts` reads. Renaming one of
 *  these breaks the listener silently, so they are stated once. */
export const LAN_HTTPS_CERT_ENV = 'theDAW_HTTPS_CERT'
export const LAN_HTTPS_KEY_ENV = 'theDAW_HTTPS_KEY'
export const LAN_HTTPS_PORT_ENV = 'theDAW_HTTPS_PORT'

const LAUNCH_TOKEN_ENV = 'THEDAW_LAUNCH_TOKEN'

function nonEmpty(value: unknown): string | null {
  return typeof value === 'string' && value.trim() !== '' ? value : null
}

/**
 * The plan in a child's stdout, or null when there isn't a usable one.
 *
 * Read from the LAST line that parses, because Python's logging writes to
 * stderr but a wrapper (uv, a venv activation script, a warning filter) can
 * still put a line of its own on stdout ahead of ours.
 *
 * Null means "no plan": no listener, and the caller says so. An `enabled` plan
 * missing its url, certificate, key or vite binary is a broken contract, not a
 * usable plan, so it is null too -- starting vite without a certificate (or
 * starting nothing at all) would only produce a crash loop no one asked for.
 */
export function parseLanHttpsPlan(stdout: string): LanHttpsPlan | null {
  const lines = String(stdout ?? '')
    .split('\n')
    .map((line) => line.trim())
    .filter((line) => line.startsWith('{'))
  for (let i = lines.length - 1; i >= 0; i -= 1) {
    let raw: unknown
    try {
      raw = JSON.parse(lines[i])
    } catch {
      continue
    }
    if (!raw || typeof raw !== 'object' || Array.isArray(raw)) continue
    const obj = raw as Record<string, unknown>
    if (typeof obj.enabled !== 'boolean') continue
    if (!Number.isInteger(obj.port)) continue
    const port = obj.port as number
    if (port <= 0 || port >= 65536) continue
    const plan: LanHttpsPlan = {
      enabled: obj.enabled,
      port,
      url: nonEmpty(obj.url),
      cert: nonEmpty(obj.cert),
      key: nonEmpty(obj.key),
      vite: nonEmpty(obj.vite),
      reason: nonEmpty(obj.reason),
    }
    if (plan.enabled && (!plan.url || !plan.cert || !plan.key || !plan.vite)) return null
    return plan
  }
  return null
}

/** The one line the desktop log carries about the LAN listener. Matches
 *  `LanHttpsPlan.log_line()` on the Python side so the two launchers read the
 *  same in a log. */
export function lanHttpsLogLine(plan: LanHttpsPlan | null): string {
  if (!plan) return 'LAN (https): off - no plan could be read'
  if (plan.enabled && plan.url) return `LAN (https): ${plan.url}`
  return `LAN (https): off - ${plan.reason ?? 'unavailable'}`
}

/**
 * How to start the listener on this platform, from `frontend/`.
 *
 * The binary is the frontend's own `node_modules/.bin/vite`, resolved by
 * backend/lib/lan_https.py and carried in the plan. It is NOT `npx`: on
 * Windows cmd searches the current directory before PATH, and a
 * non-interactive npx with no node_modules present downloads a copy of vite
 * from the registry in the middle of a launch.
 *
 * On Windows that binary is `vite.cmd`, a batch shim: node refuses to spawn
 * one directly (EINVAL, since the 2024 argument-injection fix), so it goes
 * through `cmd /c` -- the same reason spawnBackend's Windows fallback does.
 * The path is quoted there because it can contain a space.
 */
export function lanListenerCommand(
  platform: string,
  plan: LanHttpsPlan,
): { command: string; args: string[] } {
  const vite = plan.vite
  if (!plan.enabled || !vite) {
    throw new Error('lanListenerCommand is for an enabled plan with a vite binary')
  }
  if (platform === 'win32') {
    const quoted = vite.includes(' ') ? `"${vite}"` : vite
    return { command: 'cmd', args: ['/c', [quoted, ...LISTENER_ARGS].join(' ')] }
  }
  return { command: vite, args: [...LISTENER_ARGS] }
}

/**
 * The listener's environment: `base` plus the three names it reads.
 *
 * `ENABLE_HMR` is not one of them -- it is passed through from `base` like
 * everything else, exactly as `listener_env` does on the Python side. Setting
 * it starts a watcher over the whole repository, and the desktop shell has no
 * use for one.
 *
 * The launch token is dropped under every spelling (Windows environment names
 * ignore case) even though `buildBaseEnv` has already dropped it: vite here
 * runs the frontend's own devDependencies, and third-party code must never be
 * able to send `X-TheDAW-Launch-Token` and pass as the desktop shell. Defence
 * in depth costs one loop and survives a future caller passing a different
 * base.
 */
export function lanListenerEnv(
  base: NodeJS.ProcessEnv,
  plan: LanHttpsPlan,
): NodeJS.ProcessEnv {
  if (!plan.enabled || !plan.cert || !plan.key) {
    throw new Error('lanListenerEnv is for an enabled plan with a certificate')
  }
  const env: NodeJS.ProcessEnv = { ...base }
  for (const key of Object.keys(env)) {
    if (key.toUpperCase() === LAUNCH_TOKEN_ENV) delete env[key]
  }
  env[LAN_HTTPS_CERT_ENV] = plan.cert
  env[LAN_HTTPS_KEY_ENV] = plan.key
  env[LAN_HTTPS_PORT_ENV] = String(plan.port)
  return env
}
